### ISSUE TITLE
Please support `Instrument` on types implementing `IntoFuture `

### ISSUE THREAD

**Issue by joshtriplett** [tokio-rs/tracing#2859]
Right now, instrumenting a type that implements `IntoFuture` requires explicitly calling `into_future().instrument(...)`. I'd love to have an impl of `Instrument` for any type implementing `IntoFuture`, to allow writing just `.instrument(...)`.

**Comment by hawkw** [comment]
Yeah, we should definitely add an implementation for types implementing `IntoFuture`; the only reason we don't have that already is historical: `tracing::Instrument` predates `IntoFuture` being stable in the standard library.

**Comment by TimDiekmann** [comment]
Would it be much work to change the current implementation?
I would guess it would be something like this
```rust
impl<T: IntoFuture> Future for Instrumented<T> {
    type Output = T::Output;

    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
        let (span, inner) = self.project().span_and_inner_pin_mut();
        let _enter = span.enter();
        inner.poll(cx)
    }
}
```
but if `T: IntoFuture`, `inner.poll()` is obviously not available.

**Comment by jplatte** [comment]
Not like that. The constructor signature would have to change.

**Comment by jplatte** [comment]
… except, that's not possible, at least not without a breaking change. Right now, `.instrument` is available on anything `Sized`, for this to work it would have to be restricted to anything `IntoFuture`.
